### DISCUSSION TITLE
Rocky Linux 9 build

### DISCUSSION THREAD

**Discussion by rgwilson68** [dreamworksanimation/openmoonray#114] [Announcements]
The source code has been updated in 1.3.0.0 to build with GCC version 11 and Qt 5.15.3, making it buildable in Rocky Linux 9.
There is an initial set of build instructions and scripts in /building/Rocky9 (openmoonray repo).

The Centos 7 build instructions are unchanged in this release, but the plan is to make some changes in line with the new Rocky 9 and move them to a Centos7 subdirectory, together with some more in-depth build documentation. This should be ready for the next release.

Note that there are dependency version updates in this release (see the release notes for details).

A couple of build FAQs in the meantime:

Can you build MoonRay without root permissions? 

   In theory, yes, but the difficult part is installing the packages we get from the Linux distribution. Installing with the yum/dnf package managers generally requires root access. They do have options to set an alternate install location, but the general opinion seems to be that it is difficult to get everything to work correctly if you do this. The dependencies built from source can be installed anywhere by editing the CMakeFiles.txt and presets files : this will get simpler in the next release. MoonRay itself can be installed anywhere, using the --prefix option to cmake --install.

Do you have to use git submodules and clone the repositories in a fixed directory structure?

   No, but it is recommended for most users. Each of the 18 source repositories can be cloned wherever you want and built separately using CMake -- this is how we build them at DreamWorks Animation. However, this requires building them in the correct order and configuring the CMake build of each repository to be able find the installed result of the other repositories it depends on. We use Rez 2 and the rez-build command to do this (that is what the "package.py" file in each repository is for). 
  We created the "openmoonray" repository to provide a way of cloning and building everything together as one CMake project, but if you have a preexisting system for managing packages and package dependencies you should be able to use that instead.

**Comment by rgwilson68** [comment]
Also, thanks to @panda1100 for the work on building on Rocky 9 in apptainer!

**Reply by panda1100** [reply]
@rgwilson68 
I made a PR https://github.com/dreamworksanimation/openmoonray/pull/122 for building MoonRay v1.3.0.0 container with Apptainer

I tested building MoonRay without `libcgroup` package, those are deleted from kojihub.stream.centos.org server (https://kojihub.stream.centos.org/koji/buildinfo?buildID=5055), and successfully built without it.

> Can you build MoonRay without root permissions?

with Apptainer, yes

**Comment by rgwilson68** [comment]
Yes, Arras now has a stub version of cgroup that it uses if libcgroup can't be found. You can just remove the line that tries to fetch libcgroup from kojihub if it is no longer there.

If cgroups are available, the Arras node can use them to restrict resources used by rendering. However, this functionality is disabled by default, and compiling without libcgroup should be fine for most people.

**Comment by mpjohn** [comment]
Hi there,
I've been having some trouble on Rocky9 with the latest build. It builds 100% but then in moonray_gui everything except lights are black. 
I had tested an old version before Rocky support which worked a lot better ( you could see models and textures ).

My graphics cards have a dev rating of 5 which is too low I know but does this explain the black?
The test rectangle works fine. I presume if the cards are too old then it ignores them for processing?
I've tried this in both moonray_gui and moonray

Here are some of the errors I'm getting if I go with the build command on the Rocky9 page it stalls at around 86% with a complaint about opencolorio....

gmake[3]: *** [Makefile:146: all] Error 2
gmake[2]: *** [CMakeFiles/OpenColorIO.dir/build.make:89: OpenColorIO-prefix/src/OpenColorIO-stamp/OpenColorIO-build] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:257: CMakeFiles/OpenColorIO.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2

But if I try the command from the main install instructions page targetting Rocky9 it works 100% ( but everything renders black )

Some of the errors I've gotten along the way...

Error: RdlInstancerGeometry("/scene/objects/"): Did not find any references geometry. Please make sure the "referencess" field contains at least one source references geometry


and specifically for the old cards...
Error creating denoiser: GPU too old, must be compute capability 6 or greater
Error: Denoiser: GPU too old, must be compute capability 6 or greater

I'm guessing this will work better when I upgrade the cards. 
Have I missed anything else? The gmake openio errors may be part of the issue I'm thinking?

Thanks

**Reply by panda1100** [reply]
Have you checked NVIDIA driver version and CUDA toolkit version compatibility. I'm guessing driver on your system too old. Just a hunch though.

**Reply by mpjohn** [reply]
Hi Panda, Thanks for the tip. I reinstalled Rocky9 from scratch and installed the nvidia drivers properly ( not a straightforward matter ). 
I noticed a few 404 errors though...

--2024-01-12 16:24:56--  https://kojihub.stream.centos.org/kojifiles/packages/libcgroup/0.42.2/5.el9/x86_64/libcgroup-0.42.2-5.el9.x86_64.rpm
Resolving kojihub.stream.centos.org (kojihub.stream.centos.org)... 66.187.232.100
Connecting to kojihub.stream.centos.org (kojihub.stream.centos.org)|66.187.232.100|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-01-12 16:25:01 ERROR 404: Not Found.

--2024-01-12 16:25:01--  https://kojihub.stream.centos.org/kojifiles/packages/libcgroup/0.42.2/5.el9/x86_64/libcgroup-devel-0.42.2-5.el9.x86_64.rpm
Resolving kojihub.stream.centos.org (kojihub.stream.centos.org)... 66.187.232.100
Connecting to kojihub.stream.centos.org (kojihub.stream.centos.org)|66.187.232.100|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2024-01-12 16:25:02 ERROR 404: Not Found.

Is that enough to throw everything off?

What would you recommend?

Thanks
M

**Reply by mpjohn** [reply]
I'm kinda new at this...the driver version is...
 NVIDIA-SMI 545.23.08              Driver Version: 545.23.08    CUDA Version: 12.3   

IS that too old?

I got some new errors too...

/bin/ld: CMakeFiles/ociolutimage.dir/main.cpp.o: undefined reference to symbol '_ZN16OpenImageIO_v2_310ParamValue11clear_valueEv'
/bin/ld: /usr/local/lib64/libOpenImageIO_Util.so.2.3: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
gmake[5]: *** [src/apps/ociolutimage/CMakeFiles/ociolutimage.dir/build.make:100: src/apps/ociolutimage/ociolutimage] Error 1
gmake[4]: *** [CMakeFiles/Makefile2:972: src/apps/ociolutimage/CMakeFiles/ociolutimage.dir/all] Error 2
gmake[4]: *** Waiting for unfinished jobs....


The thing is - I was able to build this before Xmas on this machine without Rocky9 support.
Do you think I need a machine upgrade just to build this?

Thanks
M

**Reply by panda1100** [reply]
Those two files is not available anymore. comment out those two just works. 

> Hi Panda, Thanks for the tip. I reinstalled Rocky9 from scratch and installed the nvidia drivers properly ( not a straightforward matter ). I noticed a few 404 errors though...
> 
> --2024-01-12 16:24:56-- https://kojihub.stream.centos.org/kojifiles/packages/libcgroup/0.42.2/5.el9/x86_64/libcgroup-0.42.2-5.el9.x86_64.rpm Resolving kojihub.stream.centos.org (kojihub.stream.centos.org)... 66.187.232.100 Connecting to kojihub.stream.centos.org (kojihub.stream.centos.org)|66.187.232.100|:443... connected. HTTP request sent, awaiting response... 404 Not Found 2024-01-12 16:25:01 ERROR 404: Not Found.
> 
> --2024-01-12 16:25:01-- https://kojihub.stream.centos.org/kojifiles/packages/libcgroup/0.42.2/5.el9/x86_64/libcgroup-devel-0.42.2-5.el9.x86_64.rpm Resolving kojihub.stream.centos.org (kojihub.stream.centos.org)... 66.187.232.100 Connecting to kojihub.stream.centos.org (kojihub.stream.centos.org)|66.187.232.100|:443... connected. HTTP request sent, awaiting response... 404 Not Found 2024-01-12 16:25:02 ERROR 404: Not Found.
> 
> Is that enough to throw everything off?
> 
> What would you recommend?
> 
> Thanks M

**Reply by panda1100** [reply]
@mpjohn I recommend you to check this official document about compatibility https://docs.nvidia.com/deploy/cuda-compatibility/

**Comment by mpjohn** [comment]
Thanks for your help panda1100.

I'm going to update the cards tomorrow. This might be a silly question but do I need to rebuild Moonray once I update the hardware? I'm hoping no.

Also I'm a bit disappointed that Moonray wasn't able to command line render with the older cards in place. Is that expected behavour?
I understand why it can't work with a card that's rated anything less than 6 at present but shouldn't it still render in non GPU mode?

Thanks
M

**Comment by mpjohn** [comment]
Also...I got these extra errors when I built everything from scratch... ( a ) should I be concerned? ( b ) does it explain black renders?

-- * Checking for dependencies...
-- *   - Missing a dependency 'Package'?
-- *     Try cmake -DPackage_ROOT=path or set environment var Package_ROOT=path
-- *     For many dependencies, we supply src/build-scripts/build_Package.bash
-- *   - To exclude an optional dependency (even if found),
-- *     -DUSE_Package=OFF or set environment var USE_Package=OFF 

-- Could NOT find Imath (missing: Imath_DIR)

CMake Warning (dev) at /installs/cmake-3.23.1-linux-x86_64/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:438 (message):
  The package name passed to `find_package_handle_standard_args` (JPEG) does
  not match the name of the calling package (JPEGTurbo).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  src/cmake/modules/FindJPEGTurbo.cmake:42 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  src/cmake/checked_find_package.cmake:127 (find_package)
  src/cmake/externalpackages.cmake:123 (checked_find_package)
  CMakeLists.txt:141 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Could NOT find JPEG (missing: JPEG_LIBRARIES JPEG_INCLUDE_DIRS) 

-- Could NOT find Ptex (missing: Ptex_DIR)

Make Warning at src/heif.imageio/CMakeLists.txt:11 (message):
  heif plugin will not be built


CMake Warning at src/jpeg2000.imageio/CMakeLists.txt:11 (message):
  Jpeg-2000 plugin will not be built


CMake Warning at src/raw.imageio/CMakeLists.txt:11 (message):
  Raw plugin will not be built

**Comment by mpjohn** [comment]
OK. I've updated to an RTX4000 card which should work and while the warning signs are gone everything apart from the test rectangle is still rendering black. 
Any ideas what might be happening?
Do I need to rebuild from scratch? Would that even help?

**Reply by mpjohn** [reply]
I'm stumped!!
I've reinstalled Rocky 9 from scratch - updated it. Installed the correct drivers for the RTX4000.
Then I rebuilt Moonray and everything works without any errors but I'm still getting black renders for the test scenes.
The geometry isn't being loaded I think?
The test rectangle and sphere work fine.
Maybe I'm doing something wrong? Attached are grabs of the black render of the coffee machine along with the nvidia drivers I have loaded. Any suggestions at all about what might be causing this issue?
Bearing in mind that this worked on this machine on Rocky9 before it was properly supported.

Thanks
M
![Screenshot from 2024-01-18 12-29-30](https://github.com/dreamworksanimation/openmoonray/assets/56960930/0de8db6d-4e95-4aac-ac24-7aaf19f717a2)
![Screenshot from 2024-01-18 12-53-20](https://github.com/dreamworksanimation/openmoonray/assets/56960930/075f1a68-053c-4485-9496-2df9c663854e)

**Reply by mpjohn** [reply]
Here's my debug file which suggests that the geometry is being loaded. Any advice on how to get this thing working would be greatly [appreciated!
[debug_output.txt](https://github.com/dreamworksanimation/openmoonray/files/13978161/debug_output.txt)

**Comment by jlanz** [comment]
Ah!   I see the issue in the first line of your debug_output.txt file:  It's how you are running the moonray command.

These example scenes are split into two files:   scene.rdla and scene.rdlb
The scene.rdla file is human readable (Lua script) and contains most of the scene description, but it is intentionally missing the large array attributes on the SceneObjects as they are too large and can make the file difficult for a human to read.

The scene.rdlb file is binary and contains the remaining array attributes -- things such as vertex data, normals, texture coordinates, etc.

When you run moonray (or moonray_gui) you must pass them both in on the command-line, for example:
moonray_gui -in scene.rdla -in scene.rdlb

You can read more about the RDL2 format here:
https://docs.openmoonray.org/getting-started/about/rdl-scene-format/#rdla-format
https://docs.openmoonray.org/getting-started/about/rdl-scene-format/#rdlb-format

Let us know how this works for you!
- Jon

**Comment by mpjohn** [comment]
That's perfect. I feel pretty stupid but hopefully someone out there can learn from my mistake!
Many thanks!!

**Comment by jlanz** [comment]
Please don't feel bad!  Congrats on getting it built and running (not an easy task!).

I think we could do a better job of documenting this.  I think it is fair to say that this is not going to be obvious to a new moonray user. I'll look for ways to improve our documentation on rendering these example scenes.

These scenes could have been entirely contained in a single RDLA file (long, but human readable), or in a single RDLB file (faster to load, smaller on disk), or split into both ascii/binary formats as I chose to do.  We developers sometimes work with the scene split this way for convenience when hand-editing the files during testing, and it didn't occur to me that it might be confusing to distribute them this way.

**Comment by mpjohn** [comment]
Thanks Jon. It's great to have it up and running!!

**Comment by mpjohn** [comment]
I've been doing some more testing and found the following errors...
Error: ImageMap("/scene/textures/0x21ad550"): (272288 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x21a7830"): (362760 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x21a7230"): (33968208 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x237c030"): (76400 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x21b42f0"): (386000 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x22d6a30"): (409960 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x22f2e40"): (1979536 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x22a5fc0"): (25697856 times) unknown oiio sample failure
Error: ImageMap("/scene/textures/0x229c680"): (15738664 times) unknown oiio sample failure

Is this an opencolorIO issue? Does anyone know how to solve it?
Thanks

![Screenshot from 2024-01-26 16-33-33](https://github.com/dreamworksanimation/openmoonray/assets/56960930/c1d1ce4e-e621-4875-a4b1-f73b5d25a0d6)

**Comment by mpjohn** [comment]
Maybe I should mention that since building Moonray I have installed Maya and Houdini - I've also attempted a few times to build Moonray against Houdini. Could any of this have messed up my original install of Moonray?

**Reply by BrianHanke** [reply]
Possibly. Maya and Houdini might have different library versions, might have put a conflicting symlink somewhere, etc.

**Comment by mpjohn** [comment]
Is CentOS just going to be easier to set up than Rocky Linux?
So far I've built Moonray on rocky 9.3 but been unable to get Houdini to work properly with it.
In fact since installing Houdini I seem to have broken my Moonray build and it won't successfully rebuild.
I'm beginning to think this stuff will just work better with CentOS...since that's what's used in production.
Anyone got any thoughts on this?

**Comment by mpjohn** [comment]
Hi There,

I've built Moonray against the Houdini libraries as described on the forums.
Surprisingly this worked without any issues...( at all! )

So now I can go into Solaris and turn on Moonray.
If the scene has nothing in it then hydra works fine and I can see feedback
in the terminal.

Now if I make a component builder. Add GEO. Then add a Dwa base material
and a collector.
It crashes every time.

Any ideas what might be happening?

Thanks
Matt

**Comment by mpjohn** [comment]
I notice that the rocky 9 install_packages.sh still have a few redundant lines...

wget https://kojihub.stream.centos.org/kojifiles/packages/libcgroup/0.42.2/5.el9/x86_64/libcgroup-0.42.2-5.el9.x86_64.rpm
wget https://kojihub.stream.centos.org/kojifiles/packages/libcgroup/0.42.2/5.el9/x86_64/libcgroup-devel-0.42.2-5.el9.x86_64.rpm
dnf install libcgroup-0.42.2-5.el9.x86_64.rpm -y
dnf install libcgroup-devel-0.42.2-5.el9.x86_64.rpm -y

...which if I'm not mistaken will cause a 404 error. It might be worth editing this for the next version?
